src/rpc/claimtrie.cpp

```diff
@@ -102,91 +107,194 @@ std::string escapeNonUtf8(const std::string& name)
+    return true;
+}
+
+bool getClaimById(const uint160& claimId, std::string& name, CClaimValue* claim = nullptr)
+{
+    if (claimId.IsNull())
+        return false;
+
+    CClaimIndexElement element;
+    if (!pclaimTrie->db->Read(std::make_pair(CLAIM_BY_ID, claimId), element))
+        return false;
+    if (element.claim.claimId == claimId) {
+        name = element.name;
+        if (claim)
+            *claim = element.claim;
         return true;
     }
     return false;
 }
 
+// name can be setted explicitly
+bool getClaimById(const std::string& partialId, std::string& name, CClaimValue* claim = nullptr)
+{
+    if (partialId.empty())
+        return false;
+
+    std::unique_ptr<CDBIterator> pcursor(pclaimTrie->db->NewIterator());
+
+    for (pcursor->SeekToFirst(); pcursor->Valid(); pcursor->Next()) {
+        std::pair<uint8_t, uint160> key;
+        if (!pcursor->GetKey(key) || key.first != CLAIM_BY_ID)
+            continue;
+
+        if (key.second.GetHex().find(partialId) != 0)
+            continue;
+
+        CClaimIndexElement element;
+        if (pcursor->GetValue(element)) {
+            if (!name.empty() && name != element.name)
+                continue;
+            name = element.name;
+            if (claim)
+                *claim = element.claim;
+            return true;
+        }
+    }
+    return false;
+}
+
+std::vector<CClaimNsupports> seqSort(const std::vector<CClaimNsupports>& source)
+{
+    auto claimsNsupports = source;
+
+    std::sort(claimsNsupports.begin(), claimsNsupports.end(), [](const CClaimNsupports& lhs, const CClaimNsupports& rhs) {
+        auto& lc = lhs.claim;
+        auto& rc = rhs.claim;
+        return lc.nHeight < rc.nHeight || (lc.nHeight == rc.nHeight && lc.outPoint.n < rc.outPoint.n);
+    });
+
+    return claimsNsupports;
+}
+
+std::size_t indexOf(const std::vector<CClaimNsupports>& source, const uint160& claimId)
+{
+    auto it = std::find_if(source.begin(), source.end(), [&claimId](const CClaimNsupports& claimNsupports) {
+        return claimNsupports.claim.claimId == claimId;
+    });
+    assert(it != source.end());
+    return std::distance(source.begin(), it);
+}
+
+UniValue claimToJSON(const CCoinsViewCache& coinsCache, const CClaimValue& claim)
+{
+    UniValue result(UniValue::VOBJ);
+
+    std::string targetName;
+    if (getClaimById(claim.claimId, targetName))
+        result.pushKV(T_NAME, escapeNonUtf8(targetName));
+
+    auto& coin = coinsCache.AccessCoin(claim.outPoint);
+    if (!coin.IsSpent()) {
+        std::string value;
+        if (extractValue(coin.out.scriptPubKey, value))
+            result.pushKV(T_VALUE, value);
+
+        CTxDestination address;
+        if (ExtractDestination(coin.out.scriptPubKey, address))
+            result.pushKV(T_ADDRESS, EncodeDestination(address));
+    }
+
+    result.pushKV(T_CLAIMID, claim.claimId.GetHex());
+    result.pushKV(T_TXID, claim.outPoint.hash.GetHex());
+    result.pushKV(T_N, (int)claim.outPoint.n);
```

Comment:
should we cast this?

Reply:
UniValue does not provide an uint32 constructor which can lead to ambiguous.